src/py4vasp/calculation/_partial_charge.py

```diff
@@ -0,0 +1,434 @@
+        return np.max(cart_coords[:, 2])
+
+    def _get_lowest_z_coord(self):
+        cart_coords = _get_sanitized_cartesian_positions(self._structure)
+        return np.min(cart_coords[:, 2])
+
+    def _topology(self):
+        return str(self._structure._topology())
+
+    def _estimate_vacuum(self):
+        _raise_error_if_vacuum_not_along_z(self._structure)
+        slab_thickness = self._get_highest_z_coord() - self._get_lowest_z_coord()
+        return self._out_of_plane_vector() - slab_thickness
+
+    def _raise_error_if_tip_too_far_away(self, tip_height):
+        if tip_height > self._estimate_vacuum() / 2:
+            message = f"""The tip position at {tip_height:.2f} is above half of the
+             estimated vacuum thickness {self._estimate_vacuum():.2f} Angstrom.
+            You would be sampling the bottom of your slab, which is not supported."""
+            raise exception.IncorrectUsage(message)
+
+    def _get_stm_data(self, spin):
+        if 0 not in self.bands() or 0 not in self.kpoints():
+            massage = """Simulated STM images are only supported for non-separated bands and k-points.
+            Please set LSEPK and LSEPB to .FALSE. in the INCAR file."""
+            raise exception.NotImplemented(massage)
+        chg = self._correct_units(self.to_numpy(spin, band=0, kpoint=0))
+        return self._smooth_stm_data(chg)
+
+    def _correct_units(self, charge_data):
+        grid_volume = np.prod(self.grid())
+        cell_volume = self._structure.volume()
+        return charge_data / (grid_volume * cell_volume)
+
+    def _smooth_stm_data(self, data):
+        sigma = (
+            self.stm_settings.sigma_xy,
+            self.stm_settings.sigma_xy,
+            self.stm_settings.sigma_z,
+        )
+        return ndimage.gaussian_filter(
+            data, sigma=sigma, truncate=self.stm_settings.truncate, mode="wrap"
+        )
+
+    def _get_stm_plane(self):
+        """Return lattice plane spanned by a and b vectors"""
+        # lv = self._structure.lattice_vectors()
+        # _raise_error_if_vacuum_not_along_z(self._structure)
+        # l0 = np.linalg.norm(lv[0]) / np.linalg.norm(lv[0,:2]) * lv[0,:2]
+        # l1 = np.linalg.norm(lv[1]) / np.linalg.norm(lv[1,:2]) * lv[1,:2]
+        # from py4vasp._util.slicing import Plane
+        # return Plane(vectors=np.vstack((l0, l1)),
+        #             cell=lv,
+        #             cut="c",
+        #             )
+        return plane(
+            cell=self._structure.lattice_vectors(),
+            cut="c",
+            normal="z",
+        )
+
+    def _out_of_plane_vector(self):
+        """Return out-of-plane component of lattice vectors."""
+        lattice_vectors = self._structure.lattice_vectors()
+        _raise_error_if_vacuum_not_along_z(self._structure)
+        return lattice_vectors[2, 2]
+
+    def _spin_polarized(self):
+        return self._raw_data.partial_charge.shape[2] == 2
```

Comment:
Did you think of noncollinear densities?

Reply:
Noncollinear partial charge calculations are not yet implemented in VASP.